config.go

```diff
@@ -0,0 +1,324 @@
+// go-config - Library for reading cacophony config files.
+// Copyright (C) 2018, The Cacophony Project
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+package config
+
+import (
+	"context"
+	"errors"
+	"path"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/gofrs/flock"
+	"github.com/mitchellh/mapstructure"
+	"github.com/spf13/afero"
+	"github.com/spf13/viper"
+)
+
+type Config struct {
+	v                *viper.Viper
+	fileLock         *flock.Flock
+	accessedSections []string //TODO record each section accessed for restarting service purpose
```

Comment:
This should probably be a "set" to avoid duplicate entries. You can mimic a set with a `map[string]struct{}`